backend/bitboxbase/updater/updater.go

```diff
@@ -83,7 +108,150 @@ func (updater *Updater) Connect(address string, bitboxBaseID string) error {
+	keypair := updater.configGetAppNoiseStaticKeypair()
+	if keypair == nil {
+		updater.log.Info("noise static keypair created")
+		kp, err := cipherSuite.GenerateKeypair(rand.Reader)
+		if err != nil {
+			return errors.New("unable to generate a new noise keypair for the wallet app communication with the BitBox Base")
+		}
+		keypair = &kp
+		if err := updater.configSetAppNoiseStaticKeypair(keypair); err != nil {
+			updater.log.WithError(err).Error("could not store app noise static keypair")
+			// Not a critical error, ignore.
+		}
+	}
+	handshake, err := noise.NewHandshakeState(noise.Config{
+		CipherSuite:   cipherSuite,
+		Random:        rand.Reader,
+		Pattern:       noise.HandshakeXX,
+		StaticKeypair: *keypair,
+		Prologue:      []byte("Noise_XX_25519_ChaChaPoly_SHA256"),
+		Initiator:     true,
+	})
+	if err != nil {
+		return errors.New("failed to generate a new noise handshake state for the wallet app communication with the BitBox Base")
+	}
+	//Check the websocket connection
+	err = client.WriteMessage(1, []byte(opICanHasHandShaek))
+	if err != nil {
+		return errors.New("unable to write BitBox Base Handshake request to websocket")
+	}
+	_, responseBytes, err := client.ReadMessage()
+	responseSuccess := []byte("ACK")
+	if err != nil {
+		return errors.New("unable to read BitBox Base Handshake response from websocket")
+	}
+	if string(responseBytes) != string(responseSuccess) {
+		return errors.New("no ACK received from BitBox Base as response to hanshake request")
+	}
+
+	// Do 3 part noise 'XX' handshake.
+	msg, _, _, err := handshake.WriteMessage(nil, nil)
+	if err != nil {
+		return errors.New("noise failed to write the first handshake message")
+	}
+	err = client.WriteMessage(websocket.TextMessage, msg)
+	if err != nil {
+		return errors.New("the websocket failed writing the first noise handshake message")
+	}
+	_, responseBytes, err = client.ReadMessage()
+	if err != nil {
+		return errors.New("the websocket failed reading the second noise handshake message")
+	}
+	_, _, _, err = handshake.ReadMessage(nil, responseBytes)
+	if err != nil {
+		return errors.New("noise failed to read the second handshake message")
+	}
+	msg, updater.receiveCipher, updater.sendCipher, err = handshake.WriteMessage(nil, nil)
+	if err != nil {
+		return errors.New("noise failed to write the third handshake message")
+	}
+	err = client.WriteMessage(websocket.TextMessage, msg)
+	if err != nil {
+		return errors.New("the websocket failed writing the third handshake message")
+	}
+
+	// Check if the user already authenticated the channel binding hash
+	updater.baseNoiseStaticPubkey = handshake.PeerStatic()
+	pairingVerificationRequiredByApp := !updater.configContainsBitBoxBaseStaticPubkey(
+		updater.baseNoiseStaticPubkey)
+	if pairingVerificationRequiredByApp {
+		msg = []byte(opICanHasPairinVerificashun)
+	}
+	err = client.WriteMessage(websocket.TextMessage, updater.sendCipher.Encrypt(nil, nil, msg))
```

Comment:
it's a great idea to encrypt those messages, but imho it is better to stick to the same protocol as bitbox02.

deviating from the bitbox02 protocol means it's harder to refactor noise into a package which both can use. i think it's better to stick to the same thing, as it's quite a lot of duplicated code that can be abstracted and reused.

Reply:
I'd like to make at least one normally encrypted message as part of the handshake though, since it reveals whether you assigned the sendCipher and receiveCipher in the correct order on both sides of the handshake. I don't think that we can unify the two handshake functions soon in a manner that makes sense, since the bitbox02 requires some extra state to carry around and has a completely different communication protocol.

Reply:
> I'd like to make at least one normally encrypted message as part of the handshake though, since it reveals whether you assigned the sendCipher and receiveCipher in the correct order on both sides of the handshake. 

I see. You can achieve the same with unit tests or manual tests. Once the two request/respond message pairs have been exchanged, sendCipher and receiveCipher are set and good to go (it can't fail). I wouldn't deviate from the bitbox02 handshake protocol because of this, it.

> I don't think that we can unify the two handshake functions soon in a manner that makes sense, since the bitbox02 requires some extra state to carry around and has a completely different communication protocol.

The noise handshake protocol is (can be / should be) the same though, and is like 90% there already. The protocol inside (encrypted, domain specific) does not matter for this abstraction layer.

When refactored, on the initiator (app) side it could be as easy as `noisexx.Handshake(conn, confirmPairingHashCallback)` (with type `Handshake: func(io.ReadWriter, confirmPairingHashCallback string) bool)` or similar) in the beginning, taking care of all the details, used by both the base and bitbox02 parts.

Reply:
Unit tests could be used on the desktop app side, but the desktop app has no guarantee that the other party has it correct. Obviously the other party can make unit tests as well, but it sounds preferable to me to have a round of communication that ensures that both parties have it setup correctly. If the setup happens to be incorrect on the base side and it tries to encrypt something with the `receiveCipher`, we'd only notice once we tried to decrypt something later on. From what I can gather from tls for example, having a round of test encryption is standard practice for a handshake.

Reply:
While I am not sure of the benefit, I guess it's fine to do it, but then it needs to be done correctly:

- send a random (long) challenge, which the base can echo back, and check it comes back correctly
- could do it as the first thing after the handshake (extend the handshake), which is safer and easier than trying to combine it with the previous steps

The current 'v' (or garbage) does not properly test if the encrypted channel works.

Reply:
Sounds good to me, what challenge size would you go for? Would you then also be fine with encrypting the pairing verification request? Is there a reason why the BitBox sends `"\x01"` as the pairing request and not `"v"`?

Reply:
Another key difference is how the `pairingVerificationRequiredByApp` is communicated. In short, in your case it is:
```
if pairingVerificationRequiredByDevice || pairingVerificationRequiredByApp {
    device.communication.SendFrame(opICanHasPairinVerificashun);
    reponse = device.communication.ReadFrame() 
    device.channelHashDeviceVerified = string(response) == responseSuccess
```
And in mine:
```
if pairingVerificationRequiredByApp {
    msg = []byte(opICanHasPairinVerificashun)
}
err = client.WriteMessage(websocket.BinaryMessage, msg) 
if pairingVerificationRequiredByBase || pairingVerificationRequiredByApp {
    _, responseBytes, err := client.ReadMessage()
    updater.channelHashBitBoxBaseVerified = string(responseBytes) == string(responseSuccess)
```
The main difference being that in my implementation the base listens for a request by the app even if no verification will be required, while in yours, the bitbox only receives a message if the pairing verification needs to be done.
This seems to mean that I should make the verification a separate api call for the base, such that it does not start listening indefinitely for a verification request from the wallet app that will never happen. This would require a special message in the base's api. Is it fine, if I add a TODO for now and fix this when we have a proper message protocol in place?

Reply:
On second though it seems weird to me that the app could completely skip over verification, even if the BitBox demands it, by just not sending the verification request.

Reply:
challenge: 16 or 32 bytes

> Would you then also be fine with encrypting the pairing verification request? 

As elaborated, I'd prefer to have the same protocol as bb02, so not encrypted.

>  Is there a reason why the BitBox sends "\x01" as the pairing request and not "v"?

Technically it is a response status code, not a request op code. Could have been "v" nonetheless, so no good reason ;) In bb02 the response here is basically a boolean `require_pairing_verification`.

 > On second though it seems weird to me that the app could completely skip over verification, even if the BitBox demands it, by just not sending the verification request.

The app cannot skip over it if the bitbox02 demands it. It blocks communication until it is done, if the bitbox requires it. See those lines:

https://github.com/digitalbitbox/bitbox02-firmware/blob/254e7e357c3303abf95afb8fa47ff695f062cabf/src/usb/noise.c#L244-L274

>  This would require a special message in the base's api. Is it fine, if I add a TODO for now and fix this when we have a proper message protocol in place?

The handhshake/verification protocol is kind of independent of the rest (e.g. you already send `v` instead of JSON, which is totally fine here). It's okay to do it later, but prefer to do the handshake right from the start.